File: backend/test/battle-rules/hp-pool.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { isDefeated, withDamage, withMaxIncreased } from "../../src/battle-rules/hp-pool.js";

test("isDefeated is false while current is above 0", () => {
  assert.equal(isDefeated({ max: 100, current: 1 }), false);
});

test("isDefeated is true once current reaches 0", () => {
  assert.equal(isDefeated({ max: 100, current: 0 }), true);
});

test("withDamage subtracts from current", () => {
  const pool = withDamage({ max: 100, current: 100 }, 30);
  assert.deepEqual(pool, { max: 100, current: 70 });
});

test("withDamage clamps at 0 instead of going negative", () => {
  const pool = withDamage({ max: 100, current: 20 }, 35);
  assert.deepEqual(pool, { max: 100, current: 0 });
});

test("withDamage throws for a negative amount", () => {
  assert.throws(() => withDamage({ max: 100, current: 100 }, -1));
});

test("withMaxIncreased raises both max and current by the same amount", () => {
  const pool = withMaxIncreased({ max: 100, current: 60 }, 20);
  assert.deepEqual(pool, { max: 120, current: 80 });
});

test("withMaxIncreased throws for a negative amount", () => {
  assert.throws(() => withMaxIncreased({ max: 100, current: 100 }, -1));
});

File: packages/battle_engine/test/hp_pool_test.dart
import 'package:battle_engine/battle_engine.dart';
import 'package:test/test.dart';

void main() {
  group('HpPool', () {
    test('isDefeated is false when current is above 0', () {
      const pool = HpPool(max: 100, current: 1);
      expect(pool.isDefeated, isFalse);
    });

    test('isDefeated is true when current is 0', () {
      const pool = HpPool(max: 100, current: 0);
      expect(pool.isDefeated, isTrue);
    });

    test('withDamage subtracts from current', () {
      const pool = HpPool(max: 100, current: 100);
      final next = pool.withDamage(30);
      expect(next.current, equals(70));
      expect(next.max, equals(100));
    });

    test('withDamage clamps current at 0, never negative', () {
      const pool = HpPool(max: 100, current: 10);
      final next = pool.withDamage(999);
      expect(next.current, equals(0));
    });

    test('withDamage rejects a negative amount', () {
      const pool = HpPool(max: 100, current: 100);
      expect(() => pool.withDamage(-1), throwsArgumentError);
    });

    test('withMaxIncreased raises both max and current', () {
      const pool = HpPool(max: 100, current: 60);
      final next = pool.withMaxIncreased(20);
      expect(next.max, equals(120));
      expect(next.current, equals(80));
    });

    test('withMaxIncreased rejects a negative amount', () {
      const pool = HpPool(max: 100, current: 100);
      expect(() => pool.withMaxIncreased(-1), throwsArgumentError);
    });
  });
}
